lib/utils/log/file_writer.go

```diff
@@ -0,0 +1,162 @@
+// This is to allow the File reopen required by logrotate and similar tools.
+type FileSharedWriter struct {
+	logFileName string
+	fileFlag    int
+	fileMode    fs.FileMode
+	file        *os.File
+	watcher     *fsnotify.Watcher
+	closed      bool
+
+	lock sync.Mutex
+}
+
+// NewFileSharedWriter wraps the provided [os.File] in a writer that is thread safe,
+// with ability to enable filesystem notification watch and reopen file on specific events.
+func NewFileSharedWriter(logFileName string, flag int, mode fs.FileMode) (*FileSharedWriter, error) {
+	if logFileName == "" {
+		return nil, trace.BadParameter("log file name is not set")
+	}
+	logFile, err := os.OpenFile(logFileName, flag, mode)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to create the log file")
+	}
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &FileSharedWriter{
+		logFileName: logFileName,
+		fileFlag:    flag,
+		fileMode:    mode,
+		file:        logFile,
+		watcher:     watcher,
+	}, nil
+}
+
+// Write writes len(b) bytes from b to the File.
+func (s *FileSharedWriter) Write(b []byte) (int, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if s.closed {
+		return 0, ErrFileSharedWriterClosed
+	}
+
+	return s.file.Write(b)
+}
+
+// Reopen closes the file and opens it again using APPEND mode.
+func (s *FileSharedWriter) Reopen() error {
+	// If opening the file is locked we should not acquire a lock and blocking write.
+	file, err := os.OpenFile(s.logFileName, s.fileFlag, s.fileMode)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	s.lock.Lock()
+	if s.closed {
+		return trace.NewAggregate(ErrFileSharedWriterClosed, file.Close())
+	}
+	oldLogFile := s.file
+	s.file = file
+	s.lock.Unlock()
+
+	return trace.Wrap(oldLogFile.Close())
+}
+
+// RunWatcherReopen runs a filesystem watcher for rename/remove events to reopen the log.
+func (s *FileSharedWriter) RunWatcherReopen() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if s.closed {
+		return ErrFileSharedWriterClosed
+	}
+
+	return s.runWatcherFunc(s.Reopen)
+}
+
+// runWatcherFunc spawns goroutine with the watcher loop to consume events of renaming
+// or removing the log file to trigger the action function when event appeared.
+func (s *FileSharedWriter) runWatcherFunc(action func() error) error {
+	go func() {
+		for {
+			select {
+			case event, ok := <-s.watcher.Events:
+				if !ok {
+					return
+				}
+				if s.logFileName == event.Name && (event.Has(fsnotify.Rename) || event.Has(fsnotify.Remove)) {
+					slog.DebugContext(context.Background(), "Log file was moved/removed", "file", event.Name)
+					if err := action(); err != nil {
+						slog.ErrorContext(context.Background(), "Failed to take action", "error", err, "file", event.Name)
+						continue
+					}
+				}
+			case err, ok := <-s.watcher.Errors:
+				if !ok {
+					return
+				}
+				slog.ErrorContext(context.Background(), "Error received on logger watcher", "error", err)
```

Comment:
Nit: can we pass the context to all slog functions from the caller?

Reply:
unfortunately in configuration process (where we init FileSharedWriter) we don't have any context, I wanted to use  `slog.Error`, but it blacklisted in linter test

Reply:
I just meant you can pass it here from RunWatcherReopen, and RunWatcherReopen can in turn accept the context itself from its caller in applyLogConfig - there you can pass context.Background(), but then if we need to plumb a proper context here later if would be easier to do.

Reply:
got you, it was initially like that, then I removed within watcher closing by context cancelation

@r0mant I've made watcher always enabled, and removed this config flag